btree/btree.c

```diff
@@ -3529,21 +3559,429 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+						    &bop->bo_rec.r_key);
+			lev->l_idx = node_slot.s_idx;
+
+			if (node_level(node_slot.s_node) > 0) {
+				if (oi->i_key_found) {
+					lev->l_idx++;
+					node_slot.s_idx++;
+				}
+				node_child(&node_slot, &child_node_addr);
+
+				if (!address_in_segment(child_node_addr)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child_node_addr, P_NEXTDOWN);
+			} else {
+				if (!oi->i_key_found)
+					return P_LOCK;
+				/**
+				 * If root is an internal node and it contains
+				 * only two record, if any of the record is
+				 * going to be deleted, load the other child of
+				 * root.
+				 */
+				if (oi->i_used > 0 &&
+				    node_count_rec(oi->i_level[0].l_node) == 2)
+					return handle_root_case(bop);
+
+				return P_LOCK;
+			}
+		} else {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
+		}
+	case P_STORE_CHILD: {
+		/* store child of the root. */
+		oi->i_level[1].l_sib = oi->i_nop.no_node;
+		oi->i_level[1].l_sib_seq = oi->i_nop.no_node->n_seq;
+		/* Fall through to the next step */
+	}
+	case P_LOCK:
+		if (!locked(tree))
+			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
+					    bop->bo_arbor->t_desc, P_CHECK);
+		else
+			return P_CHECK;
+	case P_CHECK: {
+		oi->i_trial++;
+		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie) ||
+		    !child_node_check(oi)) {
+			if (oi->i_trial == MAX_TRIALS) {
+				if (bop->bo_flags & BOF_LOCKALL)
+					return fail(bop, -ETOOMANYREFS);
+				else
+					bop->bo_flags |= BOF_LOCKALL;
+			}
+			if (bop->bo_arbor->t_height != tree->t_height) {
+				/* If height has changed */
+				lock_op_unlock(tree);
+				level_cleanup(oi, bop->bo_tx);
+				return P_INIT;
+			} else {
+				/* If height is same */
+				lock_op_unlock(tree);
+				return P_DOWN;
+			}
+		}
+		/* Fall through to the next step */
+	}
+	case P_ACT: {
+		struct m0_btree_rec rec;
+		struct slot node_slot;
+		/**
+		 *  if key exists, delete the key, if there is an underflow, go
+		 *  to resolve function else return P_CLEANUP.
+		*/
+
+		if (!oi->i_key_found)
+			rec.r_flags = M0_BSC_KEY_NOT_FOUND;
+		else{
+			node_slot.s_node = lev->l_node;
+			node_slot.s_idx  = lev->l_idx;
+			node_del(node_slot.s_node, node_slot.s_idx, bop->bo_tx);
+			lev->l_node->n_skip_rec_count_check = true;
+			node_done(&node_slot, bop->bo_tx, true);
+			node_fix(node_slot.s_node, bop->bo_tx);
+
+			rec.r_flags = M0_BSC_SUCCESS;
+		}
+		int rc = bop->bo_cb.c_act(&bop->bo_cb, &rec);
+		if (rc) {
+			lock_op_unlock(tree);
+			return fail(bop, rc);
+		}
+
+		if (oi->i_key_found) {
+			if (oi->i_used == 0 ||
+			    node_count_rec(lev->l_node) > 0 ) {
+				/* No Underflow */
+				lev->l_node->n_skip_rec_count_check = false;
+				lock_op_unlock(tree);
+				return P_CLEANUP;
+			}
+			lev->l_node->n_skip_rec_count_check = false;
+			return m0_btree_del_resolve_underflow(bop);
```

Comment:
once all the entries got deleted i.e. there will be underflow, so we will go to m0_btree_del_resolve_underflow() function to resolve the underflow, it will delete the current node from its parent and set l_freenode flag as true for current node (l_freenode will help to detemine if node needs to be freed or not) once we done with resolving underflow we will go to P_FRREENODE where node_put() and node_free() will be done if l_freenode = true.the reasone of freeing node in P_FREENODE and not in function  m0_btree_del_resolve_underflow(bop) is node_free  as the node might be very old

Reply:
But the function m0_btree_del_resolve_underflow() first decrements the used_count before setting the l_freenode flag, which means he will have set the l_freenode for the parent of this leaf node and would forget to mark the l_freenode for this leaf node. Please confirm.